### PR TITLE
Exclude stubbed classes from subclasses after teardown

### PR DESCRIPTION
**Pull request by GCorbel**
This fix https://github.com/rspec/rspec-mocks/issues/1568.

Stubbed classes are excluded from parent subclasses after each spec.

The original issue : 

```
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "rspec", "3.12.0" # Activate the gem and version you are reporting the issue against.
  gem 'rspec-mocks', '3.12.6'
end

puts "Ruby version is: #{RUBY_VERSION}" # Ruby version is: 3.1.4

class Something; end

describe 'Test' do
  before(:each) do
    class A < Something; end
    stub_const('B', Class.new(Something))
  end

  it 'something' do
    puts Something.subclasses # => [B, A]
  end

  it 'something else' do
    puts Something.subclasses # => [B, B, A]
    # Only one occurence of B should be listed
  end
end
```

Now, `Something.subclasses` always return `[B, A]`.

### CODE REVIEW

File: lib/rspec/mocks.rb
Comment (pirj) [inline lib/rspec/mocks.rb:107]: This could be compacted to prevent endless growth.

File: lib/rspec/mocks.rb
Comment (pirj) [inline lib/rspec/mocks.rb:106]: Is it necessary to getobj?
There’s a potential race condition, and the weakref itself should quack just like the object itself (including the === probably to allow to remove from the list.
https://stackoverflow.com/questions/69185508/ruby-weakref-has-implicit-race-condition

**Review by pirj — COMMENTED** [review]
It’s nice that you’ve found a reliable solution.
I’m skeptical, however, to include it for everyone.
There might be some performance impact, and since this can introduce flakiness due to the GC race condition and an exception is not reassuring.

Still, this should be kept as a reference for those who experience issues with `subclasses`.

File: lib/rspec/mocks.rb
Comment (GCorbel) [reply lib/rspec/mocks.rb:106]: Yes, otherwise the substraction wasn't working.

I changed to rescue RefError.

File: lib/rspec/mocks.rb
Comment (GCorbel) [reply lib/rspec/mocks.rb]: I added `.compact`.

**Review by JonRowe — CHANGES_REQUESTED** [review]
Just placing a block on this as I don't want it merged without final say so @pirj I'm concerned about the require.

File: lib/rspec/mocks.rb
Comment (GCorbel) [reply lib/rspec/mocks.rb:107]: You mean with  ? 

```
diff --git a/lib/rspec/mocks.rb b/lib/rspec/mocks.rb
index a82055dd..ad697b59 100644
--- a/lib/rspec/mocks.rb
+++ b/lib/rspec/mocks.rb
@@ -104,11 +104,7 @@ module RSpec
     @excluded_subclasses = []
 
     def self.excluded_subclasses
-      @excluded_subclasses.select(&:weakref_alive?).map do |ref|
-        ref.__getobj__
-      rescue RefError
-        nil
-      end.compact
+      @excluded_subclasses
     end
 
     def self.exclude_subclass(constant)
@@ -117,7 +113,11 @@ module RSpec
 
     module ExcludeClassesFromSubclasses
       def subclasses
-        super - RSpec::Mocks.excluded_subclasses
+        super - RSpec::Mocks.excluded_subclasses.select(&:weakref_alive?).map do |ref|
+          ref.__getobj__
+        rescue RefError
+          nil
+        end.compact
       end
     end
     Class.prepend(ExcludeClassesFromSubclasses)
```

**Comment by GCorbel** [comment]
> I’m skeptical, however, to include it for everyone. There might be some performance impact, and since this can introduce flakiness due to the GC race condition and an exception is not reassuring.

I understand your worries. I would like to find a better solution but wasn't able. 

Maybe we can use a config as `excluded_stubbed_const_from_subclasses` or enable it by default and do a config as `thread_safe_garabage_collection` ?

**Comment by GCorbel** [comment]
The more I think, the more I'm in favor of a config flag. Maybe a warning message that say the flag have to be switched should be displayed when there is an error.

**Comment by pirj** [comment]
I support your idea of opting in for this with a flag.

**Comment by pirj** [comment]
We could conditionally require weakref if this option is on.

**Comment by GCorbel** [comment]
Should we have the option on by default ?

**Comment by pirj** [comment]
No. But we’re working in RSpec 4, and we can think about enabling it there unless performance considerations come up.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (GCorbel) [inline lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:1]: I took `RSpec::Mocks::MarshalExtension` for example.

**Review by GCorbel — COMMENTED** [review]
I refactored, added a config flag and specs. I tried to respect the existing code. Let me know what you think.

**Comment by GCorbel** [comment]
I change to run specs only for Ruby 3.1 and ahead as in https://github.com/rails/rails/blob/v7.0.8/activesupport/lib/active_support/ruby_features.rb

**Comment by GCorbel** [comment]
I fixed the code for Ruby <= 2.4

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (GCorbel) [inline lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb]: Rubocop triggers this warning : 

```
lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:47:11: W: Lint/DuplicateMethods: Method RSpec::Mocks::ExcludeStubbedClassesFromSubclasses::Class#subclasses is defined at both lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:38 and lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:47.
          alias subclasses subclasses_without_rspec_mocks
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It's wrong because the method is undefined at line 46. I disabled this warning.

**Comment by GCorbel** [comment]
The CI pass except for "ruby-head" which is not to my commit

**Comment by GCorbel** [comment]
I fixed a flap on specs. Now the CI should pass.

**Comment by GCorbel** [comment]
I finally succeeded to install Ruby 3.4-dev locally and confirm that the error on the CI is not related to the PR. It also fail on the main branch.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (pirj) [inline lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:13]: Is “something” intentional wording?

File: lib/rspec/mocks.rb
Comment (pirj) [inline lib/rspec/mocks.rb:119]: I recall we wanted to adjust our style not to align by default, to avoid churn. I’d rather have inconsistently aligned neighbour lines, but no pressure.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (pirj) [inline lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:30]: Will we keep extending on every `enable!` (intermingled with disable!)?

File: spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb
Comment (pirj) [inline spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb:18]: I don’t quite understand this. We enable the e feature, the Class is altered, and we expect the second call to enable not to … remove the method added by previous enable?

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (pirj) [inline lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:30]: On another thought, can we use this class to keep the list instead of RSpec::Mocks?

File: spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb
Comment (pirj) [inline spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb:40]: Is there a reason to write this differently from how the opposite is written below?

expect(TestClass.subclasses).to include(subclass)

**Review by pirj — COMMENTED** [review]
Please accept my apologies for the wait. The code is mostly good, thanks!

I’d love to hear what @JonRowe says.
Do we need a benchmark to measure the overhead in `reset` calls?

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (GCorbel) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb]: No, it wasn't. I changed.

File: lib/rspec/mocks.rb
Comment (GCorbel) [reply lib/rspec/mocks.rb:119]: I changed and applied the rule for other lines in this file.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (GCorbel) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb]: No, if we call `enable!` multiple times, it extends only once because of `return if Class.respond_to?(:subclasses_with_rspec_mocks)`.

File: spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb
Comment (GCorbel) [reply spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb]: I agree this wasn't making sense. I changed.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (GCorbel) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb]: I agree and changed for that.

File: spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb
Comment (GCorbel) [reply spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb]: I changed to be similar to other tests.

**Comment by GCorbel** [comment]
> Please accept my apologies for the wait. The code is mostly good, thanks!

No problem, it's open source. You don't even have to reply.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (pirj) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb]: The problem I’m thinking of is when you enable!, disable!, and then enable! again when Class is not augmented, but the ExcludeStubbed… is.
Do you see what I mean?
You could add a described_class.disable! in between two enable! calls in the corresponding spec, and see how the `.once` expectation fails. I did not test this, just an assumption. I’ll be happy if it’s just my imagination.

What if we just define those two methods on the ExcludeStubbed immediately and don’t ‘extend’ it at all?

Comment (GCorbel) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb]: ok, my bad. I take a look.

Comment (GCorbel) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb]: I refactored the code. Doing and `extend` was now useless. I also changed and added missing specs.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (pirj) [inline lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:50]: This is in a few places. Might be it belongs to `enable!`, too?

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (pirj) [inline lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:35]: Ok to move this line into `enable!`?

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (pirj) [inline lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:39]: It slipped my mind, isn’t it possible to squash that so that rescue is on the map itself?

Comment (pirj) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:39]: Ah, nevermind, Ruby 1.8 probably doesn’t support that. Do you mind making a note to indicate that?

**Review by pirj — APPROVED** [review]
Looks solid, thank you! 🙌 

Just a few optional, subjective and cosmetic notes - please feel free to ignore.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (GCorbel) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:35]: I'm not sure if it's a problem but now we can do `RSpec::Mocks::ExcludeStubbedClassesFromSubclasses.excluded_subclasses` without `ExcludeStubbedClassesFromSubclasses.enable!` first. If we move this line to `enable!`, it will raise an error.

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (GCorbel) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb]: I removed the comment because the CI was failing because of that (I don't have the same behavior locally with the same version of Ruby/Rubocop).

File: lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
Comment (GCorbel) [reply lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb:35]: As you wish, I can do : 


```
diff --git a/lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb b/lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
index bb72101e..129ff0d2 100644
--- a/lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
+++ b/lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb
@@ -8,6 +8,8 @@ module RSpec
         def enable!
           return unless RUBY_VERSION >= "3.1"
           return if Class.respond_to?(:subclasses_with_rspec_mocks)
+          @excluded_subclasses = []
+          require 'weakref' unless defined?(::WeakRef)
 
           Class.class_eval do
             def subclasses_with_rspec_mocks
@@ -32,9 +34,6 @@ module RSpec
         end
 
         def excluded_subclasses
-          require 'weakref' unless defined?(::WeakRef)
-
-          @excluded_subclasses ||= []
           @excluded_subclasses.select(&:weakref_alive?).map do |ref|
             begin
               ref.__getobj__
@@ -45,9 +44,6 @@ module RSpec
         end
 
         def exclude_subclass(constant)
-          require 'weakref' unless defined?(::WeakRef)
-
-          @excluded_subclasses ||= []
           @excluded_subclasses << ::WeakRef.new(constant)
         end
       end
diff --git a/spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb b/spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb
index 3b5ded94..16454922 100644
--- a/spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb
+++ b/spec/rspec/mocks/exclude_stubbed_classes_from_subclasses_spec.rb
@@ -63,6 +63,10 @@ if RUBY_VERSION >= '3.1'
         end
 
         describe '.excluded_subclasses' do
+          before do
+            described_class.enable!
+          end
+
           it 'returns excluded subclasses' do
             subclass = Class.new
             described_class.exclude_subclass(subclass)
@@ -75,14 +79,12 @@ if RUBY_VERSION >= '3.1'
             described_class.exclude_subclass(subclass)
 
             subclass = nil
-
             GC.start
 
             expect(described_class.excluded_subclasses).to eq([])
           end
 
           it 'does not return excluded subclasses that raises a ::WeakRef::RefError' do
-            require 'weakref'
             subclass = double(:weakref_alive? => true)
             described_class.instance_variable_set(:@excluded_subclasses, [subclass])
 
```

**Comment by GCorbel** [comment]
@JonRowe can you give a look at this request ?

**Comment by JonRowe** [comment]
I'd like to apologise for leaving this hanging for so long, its been on my review list for ages but I just haven't been able to get around to it, I had a couple of concerns about weak ref and other such behaviour but following on from the discussion in #1568 I'm going to close this. Sorry to have wasted your time.

**Pull request closed without merging**
